File: 875.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        long long int rit = *max_element(piles.begin(), piles.end());
        long long int left = 1;
        long long int m = 1; // eq to k
        long long int tot = 0;
        long long int sol = INT_MAX;
        while(left <= rit){
            m = left + (rit - left)/2;
            tot = 0;
            for(int i = 0; i < piles.size(); i++){
                tot += piles[i]/m;
                if(piles[i]%m != 0)tot++;
            }
            cout<<"Tot: "<<tot<<endl;
            cout<<"Left: "<<left<<" rit: "<<rit<<" m: "<<m<<endl;
            if(tot > h){
                left = m+1;
            }else{
                rit = m-1;
                sol = min(m,sol);
            }
        }
        return sol;
    }
};

File: Python/875.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def condition(val):
            cnt = 0
            for i in range(len(piles)):
                cnt += piles[i]//val
                if piles[i]%val:
                    cnt+=1
                if cnt > h:
                    return False
            return True

        l = 1
        r = max(piles)

        while l < r:
            m = l + (r-l)//2
            if condition(m):
                r = m
            else:
                l = m+1
        return l
